test/bridge/limiter/RateLimitedBridge.sol

```diff
@@ -0,0 +1,471 @@
+        uint256 fee,
+        bytes32 kappa
+    ) external;
+
+    function retryWithdrawAndRemove(
+        address to,
+        IERC20 token,
+        uint256 amount,
+        uint256 fee,
+        address pool,
+        uint8 swapTokenIndex,
+        uint256 swapMinAmount,
+        uint256 swapDeadline,
+        bytes32 kappa
+    ) external;
+
+    function withdrawAndRemove(
+        address to,
+        IERC20 token,
+        uint256 amount,
+        uint256 fee,
+        address pool,
+        uint8 swapTokenIndex,
+        uint256 swapMinAmount,
+        uint256 swapDeadline,
+        bytes32 kappa
+    ) external;
+}
+
+interface ISynapseERC20 {
+    function MINTER_ROLE() external view returns (bytes32);
+
+    function grantRole(bytes32 role, address account) external;
+
+    function revokeRole(bytes32 role, address account) external;
+}
+
+contract RateLimitedBridge is Test {
+    address public immutable bridge;
+
+    address public constant NODE_GROUP =
+        0x230A1AC45690B9Ae1176389434610B9526d2f21b;
+
+    Utilities internal immutable utils;
+
+    RateLimiter internal immutable rateLimiter;
+
+    address payable public immutable attacker;
+    address payable public immutable limiter;
+    address payable public immutable user;
+    address payable public immutable governance;
+
+    constructor(address _bridge) {
+        bridge = _bridge;
+
+        utils = new Utilities();
+        address impl = deployCode(
+            "./artifacts/SynapseBridge.sol/SynapseBridge.json"
+        );
+        utils.upgradeTo(_bridge, impl);
+
+        rateLimiter = new RateLimiter();
+        rateLimiter.initialize();
+
+        address payable[] memory users = utils.createUsers(10);
+        attacker = users[0];
+        limiter = users[1];
+        user = users[2];
+        governance = users[3];
+    }
+
+    function setUp() public {
+        // Grant RateLimiter roles
+        rateLimiter.grantRole(rateLimiter.LIMITER_ROLE(), limiter);
+        rateLimiter.grantRole(rateLimiter.BRIDGE_ROLE(), bridge);
+        rateLimiter.grantRole(rateLimiter.GOVERNANCE_ROLE(), governance);
+
+        // Set RateLimiter address in Bridge
+        bytes32 role = IBridge(bridge).GOVERNANCE_ROLE();
+        startHoax(utils.getRoleMember(bridge, role));
+        IBridge(bridge).setRateLimiter(address(rateLimiter));
+        IBridge(bridge).setRateLimiterEnabled(true);
+        vm.stopPrank();
+
+        // Grant RateLimiter needed Bridge role
+        role = IBridge(bridge).RATE_LIMITER_ROLE();
+        hoax(utils.getRoleMember(bridge, 0x00));
+        IBridge(bridge).grantRole(role, address(rateLimiter));
+
+        // Set Bridge address in RateLimiter
+        hoax(governance);
+        rateLimiter.setBridgeAddress(bridge);
+    }
+
+    /**
+     * @dev Checks that kappaMap isn't rekt by the implementation upgrade,
+     * also checks that bridgeVersion is updated
+     */
+    function _testUpgrade(bytes32[] memory kappas) internal {
```

Comment:
Should we be maintaining this in both HH Upgradeable.ts and in Foundry? Do we want to pick where to test upgradeability & storage slots? Depending on which one to focus on (or both?), should test additional storage slot ordering.